### PR TITLE
tokenize_midi_dataset reproducing source file tree, overwrite_mode

### PR DESCRIPTION
**Pull request by Natooz**
Following #79, this PR make `tokenize_midi_dataset` reproduce the source files file tree in `out_dir`.
It also allows to handle file overwriting if files already exist in the saving path.

The file tree reproduction has also to be made for `data_augmentation_dataset`

### CODE REVIEW

**Comment by Natooz** [comment]
@leleogere I'll leave this open a few days in case you want to review / test 😄

File: miditok/midi_tokenizer.py [hunk 1]
```diff
@@ -1663,30 +1663,32 @@ def apply_bpe_to_dataset(
...
             self.apply_bpe(seq)
 
-            out_ = (
-                Path(out_path) / path.relative_to(dataset_path)
+            saving_path = (
+                out_path / json_path.relative_to(dataset_path)
                 if out_path is not None
-                else path
+                else json_path
             )
```
Comment (leleogere) [inline miditok/midi_tokenizer.py:1689]: After that line, I think there is a `saving_path.parent.mkdir(parents=True, exist_ok=True)` missing as I will raise an error if the directory structure does not exist before writing the file.

File: miditok/midi_tokenizer.py [hunk 2]
```diff
@@ -1757,8 +1767,43 @@ def tokenize_midi_dataset(
         """
         out_dir = Path(out_dir)
         out_dir.mkdir(parents=True, exist_ok=True)
+
+        # User gave a path to a directory, we'll scan it to find MIDI files
+        if not isinstance(midi_paths, list):
+            if isinstance(midi_paths, str):
+                midi_paths = Path(midi_paths)
+            root_dir = midi_paths
+            midi_paths = sum(list(midi_paths.glob(f"**/*.{ext}")) for ext in MIDI_FILES_EXTENSIONS)
```
[the remark this answers is not in the document]
Comment (Natooz) [reply miditok/midi_tokenizer.py]: Well found thank you!
For the extension we should just remove the dot in the line above.
Do you want to make the code change suggestion or do I make the changes myself ?

**Comment by Natooz** [comment]
Thank you for the review, waiting for your response.
In the meantime I'll fix what has been broken here during the day

File: miditok/midi_tokenizer.py (hunk 1 above)
Comment (leleogere) [reply miditok/midi_tokenizer.py:1689]:
```suggestion
            )
            saving_path.parent.mkdir(parents=True, exist_ok=True)
```

File: miditok/midi_tokenizer.py [hunk 3]
```diff
@@ -1757,8 +1769,53 @@ def tokenize_midi_dataset(
...
         out_dir = Path(out_dir)
         out_dir.mkdir(parents=True, exist_ok=True)
+
+        # User gave a path to a directory, we'll scan it to find MIDI files
+        if not isinstance(midi_paths, list):
+            if isinstance(midi_paths, str):
+                midi_paths = Path(midi_paths)
+            root_dir = midi_paths
+            midi_paths = sum(
+                list(midi_paths.glob(f"**/*.{ext}")) for ext in MIDI_FILES_EXTENSIONS
```
Comment (leleogere) [inline miditok/midi_tokenizer.py:1779]:
```suggestion
                (list(midi_paths.glob(f"**/*.{ext}")) for ext in MIDI_FILES_EXTENSIONS),
                []
```

**Review by leleogere — COMMENTED** [review]
Suggest proposed changes (Now that I know how to do a code suggestion, I'll do that directly next time!)

**Comment by Natooz** [comment]
Committed!
Yes, it allows you to also be author of the changes, as you are the one contributing :)

**Comment by Natooz** [comment]
I'll try to improve coverage before merging

**Pull request merged**